### ISSUE TITLE
Issues applying WebTransport auth in a browser-based client

### ISSUE THREAD

**Issue by danopia** [Doridian/wsvpn#126]
👋🏻  Hi! I'm creating a Chromebook extension which provides a lightweight personal VPN client over `WebTransport`.

📈  I was able to put together a functional WebTransport VPN extension in less than 50 lines of code by importing the wsvpn-js client from NPM, which is pretty awesome and shows the power of leveraging web tech :rocket:

📉  Unfortunately, auth is a problem. I tried htpasswd and mTLS, and in both cases Chrome doesn't seem to pass the auth thru to the `WebTransport`, even if it's created on a webpage which was already authenticated with wsvpn. The errors are vague and unclear.

❓  My questions are:
1. Have you seen either authenticator mode function from a browser's WebTransport before?
2. Would it make sense to add an auth command to the protocol that replicates the existing htpasswd auth database? I could prototype this in a fork, but wanted some feedback first.

Also, WebSocket doesn't seem to have these issues. In WebSocket mode I just need to select a client certificate in an arbitrary browser tab, and then I can connect to wsvpn from the chromebook VPN menu without issue. But I want `WebTransport` :)

---

### with htpasswd
server prints: `Client failed authenticator challenge`
browser prints:

```
Failed to establish a connection to https://username:<redacted>@vpn.danopia.net/: net::ERR_METHOD_NOT_SUPPORTED.
Uncaught WebTransportError: Opening handshake failed.
```

### with mTLS

server prints nothing! it seems `new WebTransport()` doesn't get far enough to hit a wsvpn log when wsvpn asks for a client cert.
browser prints:

```
Failed to establish a connection to https://vpn.danopia.net/: net::ERR_QUIC_PROTOCOL_ERROR.QUIC_IETF_GQUIC_ERROR_MISSING ().
Uncaught WebTransportError: Opening handshake failed.
```

**Comment by Doridian** [comment]
I have not tried WebTransport with authentication in a browser, yet. I will poke around to see if there is any ability to send authentication in such a manner.

I know it works with the binary client, so it is possible at the protocol level, but, yeah.

If there really is no way, I would be happy to take an in-band auth mechanism (I know even for WebSocket, headers are discouraged from usage, so this would be good for getting WSVPN more into "compliance") that could be applied to both WebSocket and WebTransport as a PR.

**Comment by Doridian** [comment]
Just found this issue on the standard: https://github.com/w3c/webtransport/issues/263 (This also provides a possible work-around for auth: Stick it in the URL and make the auth plugin read URL path instead. I **do not** like that approach, in-band would be preferable for sure)

**Comment by danopia** [comment]
Hi again! Thanks for your responses and finding the additional context :)

> I know it works with the binary client, so it is possible at the protocol level, but, yeah.

Yea, this is a Chrome/browser limitation here. Especially with mTLS as that's not even something the WebTransport protocol should be aware of AFAIK. Also, both auth modes work with WebSocket, which has about the same bootstrap from Javascript.

> Just found this issue on the standard: https://github.com/w3c/webtransport/issues/263

It's nice to see openness here, the WebSocket spec didn't really want to allow e.g. arbitrary headers when I checked in over there. Time will tell :)

> Stick it in the URL and make the auth plugin read URL path instead. I do not like that approach for several reasons

Agree, especially with something as potentially sensitive as VPN auth, and with passwords that don't expire.

> I would be happy to take an in-band auth mechanism

Nice, it sounds like there's a potential path forward to make wsvpn clients more browser-ready.

---

Unfortunately my interest in using wsvpn from a Chrome extension got blocked by [Chrome Extension Manifest V3](https://developer.chrome.com/docs/extensions/mv3/intro/)'s restrictions. There's a new timer in MV3 that shuts down extensions after 30 seconds of no user interaction, and apparently active VPN traffic doesn't count as user activity (😕). So as soon as I closed Devtools I realized the VPN only works for 30 seconds at a time and then gets killed. Doesn't matter WebSocket or WebTransport.

This 30-second limitation might get fixed at some point.. [Relax the fixed maximum lifetime for extension service workers](https://developer.chrome.com/docs/extensions/mv3/known-issues/#sw-fixed-lifetime)

In the meantime I set up a Wireguard VPN instead.. I'll close out my request here.
If I come back I'll be able to skip right to prototyping that auth command.
